src/Sylius/Component/Core/Promotion/Action/DiscountAction.php

```diff
@@ -51,34 +44,46 @@ public function __construct(FactoryInterface $adjustmentFactory, OriginatorInter
      */
     public function revert(PromotionSubjectInterface $subject, array $configuration, PromotionInterface $promotion)
     {
-        if (!$subject instanceof OrderInterface && !$subject instanceof OrderItemInterface) {
-            throw new UnexpectedTypeException(
-                $subject,
-                'Sylius\Component\Core\Model\OrderInterface or Sylius\Component\Core\Model\OrderItemInterface'
-            );
+        if (!$this->isSubjectValid($subject)) {
+            return;
         }
 
-        foreach ($subject->getAdjustments(AdjustmentInterface::ORDER_PROMOTION_ADJUSTMENT) as $adjustment) {
-            if ($promotion === $this->originator->getOrigin($adjustment)) {
-                $subject->removeAdjustment($adjustment);
+        foreach ($subject->getItems() as $item) {
+            foreach ($item->getUnits() as $unit) {
```

Comment:
I also don't like nested loops, but I'm not sure is it good for readability to extract one more function here. Making ``function``, which contains ``loop``, which contains ``function``, which contains ``loop``, which contains ``function``, which contains ``loop`` which contains a main logic doesn't seem as better way to do that 😄